### PR TITLE
User caching, some fixes

### PR DESCRIPTION
**Pull request by stappler**
!!! First review and accept pull request #2 

There are still some autogenerated files :( I am especially confused by the huge change of project.json.lock

Here are the changes:
- Added an over-engineered solution to user hashing (hashing itself is a simple generic - might need to consider hashing DLs on the server side)
- Added a couple of UTs for new classes
- Fixed a bug with null or "" named queues
- Fixed an assert thrown while logging queues with very short messages

### CODE REVIEW

**Comment by stappler** [comment]
I think I've screwed up some dependencies while rewriting the history. Because of the problem with Lan's phone I actually had to re-visit the code, which was actually really good. I've pushed another commit, which, I think, fixes server issues.

**Comment by stappler** [comment]
This pull request is cleaner than the original #2, so we are switching to using this one.

Code still contains several auto-generated files.
1) Several of them were created by VS to allow referencing the main project from test .xproj project. These are mostly contained in server/wrap folder, but there are also other ones (e.g. project.lock.json is so big, that Git doesn't even show diff for it)
2) .csproj was updated with new dependencies and new classes
3) .xproj was generated by VS from manually updated Q3Server.Tests/project.json

Other notes:
1) .pubxml is not really required, but it is handy to have checked in
2) I've added several getter classes. It does separate the logic, but it might be a bit confusing. Please, check that it is a reasonable design.

File: server/src/Q3Server/Startup.cs [hunk 1]
```diff
@@ -25,13 +29,24 @@ public void Configuration(IAppBuilder app)
 
             var manager = new QManager(new QEventsListener(GlobalHost.ConnectionManager));
 
+            var userGetterCombined = new ObjectGetterTernary<User>(new UserGetterDomain(),
```
Comment (tmct) [inline server/src/Q3Server/Startup.cs]: I don't think that this ObjectGetterTernary class is justified in its generality I'm afraid - we're only using it once, and the condition is "does the first getter return null?" :) Please could you change this to e.g. just be a UserGetter class, still inheriting from IObjectGetter<User>, but explicitly containing the logic of "try Domain then Serialized"? Thanks

**Comment by tmct** [comment]
It looks like you've not made any changes to the client - is this deliberate? (If so then great: non-breaking change! :) )

File: server/src/Q3Server/GroupGetterDomain.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+                searcher.PropertiesToLoad.Add("cn");
+                searcher.PropertiesToLoad.Add("distinguishedName");
+                searcher.PropertiesToLoad.Add("mail");
+
+                var userGroups =
+                    searcher.FindAll()
+                        .Cast<SearchResult>()
+                        .Where(r => r.Properties["mail"].Count > 0)
+                        .Select(r => new { Name = (string)r.Properties["cn"][0], Path = (string)r.Properties["distinguishedName"][0] })
+                        .Where(r => !r.Path.ToLowerInvariant().Contains("security groups") || r.Name.StartsWith("Softwire - ") || r.Name.StartsWith("Office - "))
```
Comment (tmct) [inline server/src/Q3Server/GroupGetterDomain.cs:29]: Ideally we'd put the list of DLs we want in some sort of configuration file, but this is fine for now

**Pull request merged**